source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,423 @@
+ * even one time hash searching.
+ *
+ * This is useful when some keys are frequently used and the keys are known at compile time or
+ * bootstrap time. You can get superior performance by using the inline handle. For example, the
+ * filter state always uses the filter name as the key and the filter name is known at compile time.
+ * By using the inline handle, the filter state could get the key/value pair without any hash.
+ *
+ * This class also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ *
+ * These is usage example:
+ *
+ *   // Define registry to store the inline keys. The registry should have lifetime that no shorter
+ *   // than all inline maps created by the registry.
+ *   InlineMapRegistry registry;
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
+ *   // do this by define a global variable.
+ *   InlineMapRegistry<std::string>::InlineHandle inline_handle =
+ *     registry.registerInlineKey("inline_key");
+ *
+ *   // Finalize the registry. No further changes are allowed to the registry after this point.
+ *   registry.finalize();
+ *
+ *   // Create the inline map.
+ *   InlineMapPtr inline_map = registry.createInlineMap<std::string>();
+ *
+ *   // Get value by inline handle.
+ *   inline_map->insert(inline_handle, std::make_unique<std::string>("value"));
+ *   EXPECT_EQ(*inline_map->lookup(inline_handle), "value");
+ */
+template <class Key> class InlineMapRegistry {
+public:
+  /**
+   * This is the handle used to access the key/value pair in the inline map. The handle is
+   * lightweight and could be copied and passed by value.
+   */
+  class InlineKey {
+  public:
+    /**
+     * Get the id of the inline entry in the inline array. This could be used to access the
+     * key/value pair in the inline map without hash searching.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+  private:
+    friend class InlineMapRegistry;
+
+    // This constructor should only be called by InlineMapRegistry.
+    InlineKey(uint64_t inline_id) : inline_id_(inline_id) {}
+
+    uint64_t inline_id_{};
+
+#ifndef NDEBUG
+  public:
+    // This is the registry id that the inline key created by. This is used to validate the
+    // inline key validity.
+    uint64_t registryId() const { return registry_id_; }
+
+  private:
+    InlineKey(uint64_t inline_id, uint64_t registry_id)
+        : inline_id_(inline_id), registry_id_(registry_id) {}
+    uint64_t registry_id_{};
+#endif
+  };
+
+  using RegistrationMap = absl::flat_hash_map<Key, uint64_t>;
+  using RegistrationSet = std::vector<Key>;
+
+  template <class T> class InlineMap : public InlineStorage {
+  public:
+    using TPtr = std::unique_ptr<T>;
+    using UnderlayHashMap = absl::flat_hash_map<Key, TPtr>;
```

Comment:
do we care about map footprint?

If the thing you want to put into a map is already a pointer this seems inefficient; you'll wind up with a pointer to a pointer for every entry.

In my sketch I hold the values directly. If the callers decides they need to add indirection to make moving cheaper they can always specify `InlineMap<Key, std::unique_ptr<Value>>` when they instantiate it. I think the philosophy I'm looking for here is to let the caller decide rather than forcing a level of indirection in the impl they didn't ask for

Reply:
Because I want to use inline storage to store pointer and use pointer to determine if a position is set or not.

But you are write, this is inefficient. I can refactor this.

Reply:
You can use a vector<bool> to indicate population of inline values, which is generally specialized by STL to give you 64 bools in 64 bits. This is what I did in the sketch in the comments.

Reply:
I know why I only accept unique ptr before. `flexible_array_has_nontrivial_dtor`. The variable array could onlyn be used to store objects with trivial dctor. orz. So we cannot declare a variable array like this Value inline_entries_[]; .

Another problem is we will pre-allocate memory for inline entries. If we store the value in the variable arrary directly, we may pre-allocate too much memory, esp if the size of Value type is large and the inline keys is used in discrete way.

Reply:
finally, a list is used to store the value and the inline array only store the pointer to the list node.